mono/utils/mono-threads-posix.c

```diff
@@ -43,10 +43,19 @@ extern int tkill (pid_t tid, int signal);
 
 #include <pthread.h>
 
+#include <sys/mman.h>
+#include <limits.h>    /* for PAGESIZE */
+#ifndef PAGESIZE
+#define PAGESIZE 4096
+#endif
+
 #ifdef HAVE_SYS_RESOURCE_H
 #include <sys/resource.h>
 #endif
 
+static pthread_mutex_t memory_barrier_process_wide_mutex = PTHREAD_MUTEX_INITIALIZER;
+static uint8_t memory_barrier_process_wide_helper_page[PAGESIZE] __attribute__((aligned(PAGESIZE)));
+
```

Comment:
Can this be dynamically allocated when first used ?

Reply:
The array can be allocated dynamically if there's a way to ensure that it is page aligned. I'll look into it.

Reply:
Seems like mono already has some places with `memalign`, so it should be easy.

Reply:
You can allocate a page, by allocating two pages, and finding the page-aligned byte within it.

Some Posix not all systems have `getpagesize()`.

There is `sysconf(_SC_PAGESIZE`.

Reply:
I guess the dynamic allocation is what @vargaz asked for, but all memory is essentially dynamically allocated when first used. If you have two pages of untouched memory, then you won't page for more than one page of it, and depending on its placement, you won't pay for any of it.

On Windows you could get one static aligned page with `__declspec(4096)`. Similar might be in gcc/clang, but that would forgo dynamic discovery of page size. Or you could get compile time, and confirm at runtime.

Reply:
One also wonders, if `mmap/munmap`, of fd=-1, of one byte, would be just as well.
No need to know page size, no need to mprotect.